### DISCUSSION TITLE
EventList 의 의미

### DISCUSSION THREAD

**Discussion by yangwansu** [agistep/a-es.example#8] [General]
## EventList 

EventList는 재구성(인스턴스화) 된 Aggregate가 이후에 발생하는 Event들의 목록이다. 

이것은 Event 가 영속성을 통해 Aggregate 의 구체화을 위한 소스로 사용되기 위한 목록을 포함한다. 

이해가 필요한 부분은 Aggreagte 가 인스턴스 이후에 발생하는 Event 목록을 의미한다. 

### EventList 에 Event를 포함 하는것의 의미
Aggregate 가 Client 의 요청(Command)에 대해서 처리 했고 Aggreagte 의 상태를 변경 가능하게 되었다면 Aggregate는 EventList 에게 Event 를 저장 할 것을 요청한다. 

```mermaid
flowchart LR
    A((Aggregate)) -- occurs --> B;
    B((EventList));
```

우리는 이 때 Aggregate 별로 EventList 를 구성하지 않는다. 
독립된 제공자로써 여러 Aggregate 와 협력한다. 
이유는 Aggregate 가 EventList 를 도메인오브젝트로써 협력하기 보다 인프라로써 협력 하기를 원하기 때문에 Aggregate 가 명시적으로 EventList에 의존하는 모양새를 클라언트에게 전가하고 싶지 않아서이다.

```mermaid
flowchart LR
    A((Aggregate)) -- occurs --> B;
    A1((Aggregate)) -- occurs --> B;
    A2((Aggregate)) -- occurs --> B;
    B((EventList));
```

### EventList 에 Event 가 포함 된 상태의 의미
Aggregate 가 Client 의 요청(Command)에 대해서 처리 했고 Aggregate 의 상태를 변경 한 후 아직 Event 저장소에 Publish 하지 않았음을 의미한다.

### Event Publish to Event Store And Clear
Aggregate 는 Command 를 처리를 무리(?) 없이 완료 하면  Event 는 이것은 영속화 되어 Aggregate 의 재구성하도록 Event Store 로 전달 되어야 한다.

**Comment by yangwansu** [comment]
현재 [https://github.com/agistep/a-es.example/tree/4-event-sd-solutions](https://github.com/agistep/a-es.example/blob/7b059b1c53aa94694248504f78a665cc91c5b9e5/event-core/src/main/java/io/agistep/event/EventList.java#L6) 에서 EventList 를 확인 할 수 있다.

**Comment by yangwansu** [comment]
@LenKIM 
EventList 가 위 내용을 잘 표현하고 있는가? 

Rename 이 필요하다고 생각하나?

**Reply by LenKIM** [reply]
1개 이상의 aggregate 가 1개의 EventList 를 공유한다.
더하여, 1개 이상의 aggregate 또한 EventList 에 발생한 이벤트 사건을 담는다.(occured) 
아직 이벤트 저장소에 담긴 시점이 아니다.

rename이 고민되는 부분으로는 다음과 같습니다.
- 1개 이상의 aggregate 에서 발생한 서로 다른 aggregate에서 발생한 이벤트라는 메세지를 EventList 라는 이름이 담고 있는가요?
- 이벤트 저장소에 저장되기 전의 상태를 표현할 수있는 이름은 어떠할까요?

그래서 드는 후보군은 BufferedEventsList 라는 이름은 어떤가요?
